runtime/drive_handler.go

```diff
@@ -15,125 +15,227 @@ package main
+// of container rootfs drives. The Firecracker drives are hardcoded to be read-write
+// and have no rate limiter configuration.
+func CreateContainerStubs(
+	machineCfg *firecracker.Config,
+	jail jailer,
+	containerCount int,
+	logger *logrus.Entry,
+) (*StubDriveHandler, error) {
+	var containerStubs []*stubDrive
+	for i := 0; i < containerCount; i++ {
+		isWritable := true
+		var rateLimiter *proto.FirecrackerRateLimiter
+		stubFileName := fmt.Sprintf("ctrstub%d", i)
+
+		stubDrive, err := newStubDrive(
+			filepath.Join(jail.JailPath().RootPath(), stubFileName),
+			jail, isWritable, rateLimiter, logger)
+
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create container stub drive")
+		}
+
+		machineCfg.Drives = append(machineCfg.Drives, models.Drive{
+			DriveID:      firecracker.String(stubDrive.driveID),
+			PathOnHost:   firecracker.String(stubDrive.stubPath),
+			IsReadOnly:   firecracker.Bool(!isWritable),
+			RateLimiter:  rateLimiterFromProto(rateLimiter),
+			IsRootDevice: firecracker.Bool(false),
+		})
+		containerStubs = append(containerStubs, stubDrive)
+	}
+
+	return &StubDriveHandler{
+		freeDrives: containerStubs,
+		usedDrives: make(map[string]*stubDrive),
+	}, nil
 }
 
-// stubDrivesOpt is used to make and modify changes to the stub drives.
-type stubDrivesOpt func(stubDrives []models.Drive) error
+// StubDriveHandler manages a set of stub drives. It currently only supports reserving
+// one of the drives from its set.
+// In the future, it may be expanded to also support recycling a drive to be used again
+// for a different mount.
+type StubDriveHandler struct {
+	freeDrives []*stubDrive
+	// map of id -> stub drive being used by that task
+	usedDrives map[string]*stubDrive
+	mu         sync.Mutex
+}
 
-func newStubDriveHandler(path string, logger *logrus.Entry, count int, opts ...stubDrivesOpt) (*stubDriveHandler, error) {
-	h := stubDriveHandler{
-		RootPath: path,
-		logger:   logger,
+// Reserve pops a unused stub drive and returns a MountableStubDrive that can be
+// mounted with the provided options as the patched drive information.
+func (h *StubDriveHandler) Reserve(
+	id string,
+	hostPath string,
+	vmPath string,
+	filesystemType string,
+	options []string,
+) (MountableStubDrive, error) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if len(h.freeDrives) == 0 {
+		return nil, ErrDrivesExhausted
 	}
-	drives, err := h.createStubDrives(count)
+
+	freeDrive := h.freeDrives[0]
+	options, err := setReadWriteOptions(options, freeDrive.driveMount.IsWritable)
 	if err != nil {
 		return nil, err
 	}
 
-	for _, opt := range opts {
-		if err := opt(drives); err != nil {
-			h.logger.WithError(err).Debug("failed to apply option to stub drives")
+	h.freeDrives = h.freeDrives[1:]
+	h.usedDrives[id] = freeDrive
+	return freeDrive.withMountConfig(
+		hostPath,
+		vmPath,
+		filesystemType,
+		options,
+	), nil
+}
+
+// CreateDriveMountStubs creates a set of MountableStubDrives from the provided DriveMount configs.
```

Comment:
We may want to document that the rate limiters and permissions need to be known upfront

Reply:
Good call, fixed